File: csharp/1315.sum-of-nodes-with-even-valued-grandparent.cs
/*
 * @lc app=leetcode id=1315 lang=csharp
 *
 * [1315] Sum of Nodes with Even-Valued Grandparent
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {

    private static int SumEvenGrandparent(int g, int p, TreeNode root) {
        if (root == null) {
            return 0;
        }
        return ((g + 1) % 2) * root.val + SumEvenGrandparent(p, root.val, root.left) + SumEvenGrandparent(p, root.val, root.right);
    }

    public int SumEvenGrandparent(TreeNode root) {
        return SumEvenGrandparent(-1, -1, root);
    }
}
// @lc code=end

File: java/1315.sum-of-nodes-with-even-valued-grandparent.java
/*
 * @lc app=leetcode id=1315 lang=java
 *
 * [1315] Sum of Nodes with Even-Valued Grandparent
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int sumEvenGrandparent(TreeNode root) {
        return sumEvenGrandparent(root, -1, -1);
    }

    private static int sumEvenGrandparent(TreeNode curr, int p, int pp) {
        if (curr == null) {
            return 0;
        }
        return (pp % 2 == 0 ? curr.val : 0) + sumEvenGrandparent(curr.left, curr.val, p) + sumEvenGrandparent(curr.right, curr.val, p);
    }
}
// @lc code=end

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}
